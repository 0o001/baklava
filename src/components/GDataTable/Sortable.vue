<template>
  <GBox
    class="g-sortable-wrapper"
    :flex="true"
    @click="onClick"
  >
    <slot />
    <div class="sort-icon">
      <GIcon
        class="up"
        name="triangle"
        size="6px"
        :fill="upArrowColor"
      />
      <GIcon
        class="down rotate180"
        name="triangle"
        size="6px"
        :fill="downArrowColor"
      />
    </div>
  </GBox>
</template>

<script>
import GBox from '../GBox/GBox.vue';
import GIcon from '../GIcon/GIcon.vue';

export default {
  name: 'Sortable',
  components: {
    GBox,
    GIcon,
  },
  props: {
    sortName: {
      type: String,
      default: '',
    },
    sortType: {
      type: String,
      default: '',
    },
    name: {
      type: String,
      default: '',
    },
  },
  computed: {
    upArrowColor () {
      if (this.name === this.sortName && this.sortType === 'asc') {
        return 'var(--main-grey-500)';
      }
      return 'var(--mid-grey-500)';
    },
    downArrowColor () {
      if (this.name === this.sortName && this.sortType === 'desc') {
        return 'var(--main-grey-500)';
      }
      return 'var(--mid-grey-500)';
    },
  },
  methods: {
    onClick () {
      if (this.sortName === this.name && this.sortType === 'asc') {
        return this.$emit('sort', { name: this.name, type: 'desc' });
      }

      return this.$emit('sort', { name: this.name, type: 'asc' });
    },
  },
};
</script>

<style lang="scss" scoped>
  .g-sortable-wrapper {
    cursor: pointer;
    .sort-icon {
      display: flex;
      flex-direction: column;
      margin-left: 5px;

      .rotate180{
        transform: rotate(180deg);
        margin-top: 3px;
      }
    }
  }
</style>
