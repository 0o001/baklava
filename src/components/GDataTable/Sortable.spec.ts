import Sortable from '@/components/GDataTable/Sortable.vue';
import { shallowMount, Wrapper } from '@vue/test-utils';

describe('components/GDataTable/Sortable.vue', () => {
  let wrapper: Wrapper<Sortable & { [key: string]: any }>;
  const renderWrapper = (options?: any) => {
    wrapper = shallowMount(Sortable, {
      ...options,
    });
  };

  it('should render correctly with default params', () => {
    renderWrapper();

    expect(wrapper.element).toMatchSnapshot();
  });

  describe('up Arrow', () => {
    it('should has var(--main-grey-500) color if name, sortName are equal and sortType is asc', () => {
      renderWrapper({
        propsData: {
          name: 'name',
          sortName: 'name',
          sortType: 'asc',
        },
      });

      expect(wrapper.find('.up').attributes('fill')).toEqual('var(--main-grey-500)');
    });

    it('should has var(--mid-grey-500) color if name, sortName are equal and sortType is different from asc', () => {
      renderWrapper({
        propsData: {
          name: 'name',
          sortName: 'name',
          sortType: 'sortType',
        },
      });

      expect(wrapper.find('.up').attributes('fill')).toEqual('var(--mid-grey-500)');
    });

    it('should has var(--mid-grey-500) color if name, sortName are not equal and sortType is asc', () => {
      renderWrapper({
        propsData: {
          name: 'name',
          sortName: 'sortName',
          sortType: 'asc',
        },
      });

      expect(wrapper.find('.up').attributes('fill')).toEqual('var(--mid-grey-500)');
    });
  });

  describe('down Arrow', () => {
    it('should has var(--main-grey-500) color if name, sortName are equal and sortType is desc', () => {
      renderWrapper({
        propsData: {
          name: 'name',
          sortName: 'name',
          sortType: 'desc',
        },
      });

      expect(wrapper.find('.down').attributes('fill')).toEqual('var(--main-grey-500)');
    });

    it('should has var(--mid-grey-500) color if name, sortName are equal and sortType is different from desc', () => {
      renderWrapper({
        propsData: {
          name: 'name',
          sortName: 'name',
          sortType: 'sortType',
        },
      });

      expect(wrapper.find('.down').attributes('fill')).toEqual('var(--mid-grey-500)');
    });

    it('should has var(--mid-grey-500) color if name, sortName are not equal and sortType is desc', () => {
      renderWrapper({
        propsData: {
          name: 'name',
          sortName: 'sortName',
          sortType: 'desc',
        },
      });

      expect(wrapper.find('.down').attributes('fill')).toEqual('var(--mid-grey-500)');
    });
  });

  describe('sort click', () => {
    it('should emit asc sort type if sortName and name are not equal', async () => {
      renderWrapper({
        propsData: {
          sortName: 'sortName',
          name: 'name',
          sortType: 'asc',
        },
      });

      wrapper.vm.onClick();

      expect(wrapper.emitted().sort).toBeTruthy();
      expect(wrapper.emitted().sort?.[0]?.[0]).toEqual({ name: 'name', type: 'asc' });
    });

    it('should emit asc sort type if sortName, name are equal and sortType is desc', async () => {
      renderWrapper({
        propsData: {
          sortName: 'name',
          name: 'name',
          sortType: 'desc',
        },
      });

      wrapper.vm.onClick();

      expect(wrapper.emitted().sort).toBeTruthy();
      expect(wrapper.emitted().sort?.[0]?.[0]).toEqual({ name: 'name', type: 'asc' });
    });

    it('should emit desc sort type if sortName, name are equal and sortType is asc', async () => {
      renderWrapper({
        propsData: {
          sortName: 'name',
          name: 'name',
          sortType: 'asc',
        },
      });

      wrapper.vm.onClick();

      expect(wrapper.emitted().sort).toBeTruthy();
      expect(wrapper.emitted().sort?.[0]?.[0]).toEqual({ name: 'name', type: 'desc' });
    });
  });
});
