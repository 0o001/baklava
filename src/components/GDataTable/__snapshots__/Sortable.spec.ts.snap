// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/GDataTable/Sortable.vue should render correctly with default params 1`] = `
<gbox-stub
  class="g-sortable-wrapper"
  flex="true"
>
   
  <div
    class="sort-icon"
  >
    <gicon-stub
      class="up"
      fill="var(--mid-grey-500)"
      name="triangle"
      size="6px"
      stroke="currentColor"
      strokelinecap="round"
      strokelinejoin="round"
      strokewidth="2"
      tag="i"
    />
     
    <gicon-stub
      class="down rotate180"
      fill="var(--mid-grey-500)"
      name="triangle"
      size="6px"
      stroke="currentColor"
      strokelinecap="round"
      strokelinejoin="round"
      strokewidth="2"
      tag="i"
    />
  </div>
</gbox-stub>
`;
